Update cpp pointer docs (#2643)
Codecademy/docs: content/cpp/concepts/pointers/pointers.md

```

## Null Pointers

Similar to other data types, pointers can be declared without being assigned a value.

```cpp
 int *ptr;
```

This however is a cause of concern because if the program tries to dereference and manipulate a pointer that holds no memory address, it could lead to some unexpected behavior. To prevent this, `nullptr` is used, which is a special kind of value that prevents pointers from behaving unexpectedly when they hold no memory address.

```cpp
int *ptr = nullptr; // Assigns the pointer to a null pointer value 

std::cout << *ptr << "\n"; // 0

std::string game = "Cricket";
*ptr = &game;

std::cout << *ptr << "\n"; // Cricket
```

## Pointers As Function Parameters

Pointers can also be passed into functions as arguments. This helps to manipulate values from within a function even though the value lives outside that function. 

```cpp
void changeGame(std::string *game, std::string new_game) { // defining a pointer as a function parameter
  *game = new_game; // Dereference the pointer and change the game
}

std::string game = "Cricket";
changeGame(&game, "Baseball"); // Passing a pointer as a function argument

> 

Reviewer: ```suggestion
```

  - 'paths/computer-science'
---

A **pointer** is a special type of variable that stores the memory address of another variable.

## Declaration and Initialization of Pointer variable

> A pointer variable is declared using the indirection operator (`*`). This is a special unary operator that signals to our program to store the memory address of another variable.

Reviewer 2: ```suggestion
A pointer variable is declared using the indirection operator (`*`). This is a special unary operator that signals the program to store the memory address of another variable.
```

> To get the memory address of a variable we wish to store in our pointer variable, the address operator (`&`) is used. This signals to our program to return the memory address of a variable. Memory addresses are usually prefixed with `0x` this just indicates that the follwing digits are in hexadecimal notation.

Reviewer 2: ```suggestion
To get the memory address of a variable to store in the pointer variable, the address operator (`&`) is used. This signals the program to return the memory address of a variable. Memory addresses are usually prefixed with `0x` this just indicates that the following digits are in hexadecimal notation.
```

```cpp
std::string game = "Cricket";

std::cout << game << "\n";   // Output: Cricket
std::cout << &game << "\n";  // Output of game memory address: 0x6dfed4

// A pointer with the name ptr, that stores the address of game
std::string* ptr = &game;
std::cout << ptr << "\n";    // Output of pointer to game: 0x6dfed4
```

## Pointer Dereferencing

Dereferencing a pointer means getting the value that a pointer variable is pointing at.

> We cannot perform a lot of useful operations using just the memory address that pointers store. Instead, we need a way to get the value stored in the memory address so we can manipulate it directly.

Reviewer 2: ```suggestion
There aren't many of useful operations that can be performed using just the memory address that pointers store. Instead, there needs to be a way to get the value stored in the memory address so it can be manipulated directly.
```

> Dereferencing a pointer helps us do this. We dereference a pointer with the indirection operator (`*`) as well.

Reviewer 2: ```suggestion
Dereferencing a pointer helps to do this. A pointer is derefrenced with the indirection operator (`*`) as well.
```

> Dereferencing a pointer helps to do this. A pointer is derefrenced with the indirection operator (`*`) as well.

Reviewer 3: ```suggestion
This can be done by dereferencing a pointer. The indirection operator (`*`) used for declaring a pointer can be used to dereference a pointer as well.
```

> When we dereference a pointer, we tell the program to give us the value contained in the specified memory address and then we can perform operations on it.

Reviewer 2: ```suggestion
When a pointer is dereferenced, it tells the program to return the value contained in the specified memory address and then operations can be performed on it.
```

```cpp
std::string game = "Cricket";  // Variable declaration
std::string* ptr = &game;    // Pointer declaration

std::cout << *ptr << "\n"; // Output of value dereferenced from ptr: Cricket

*ptr = "Baseball";
std::cout << *ptr << "\n"; // Output of value dereferenced from ptr: Baseball
```

## Null Poiners

> Just like with other data types, pointers can be declared without being assigned a value.

Reviewer 3: ```suggestion
Similar to other data types, pointers can be declared without being assigned a value.
```

```cpp

 int *ptr;

```

> This however is a cause of concern because if we try to dereference and manipulate a pointer that holds no memory address, it could lead to some unexpected behaviour. To prevent this we use `nullptr` which is a special kind of value that prevents our pointers from behaving unexpectedly when they hold no memory address.

Reviewer 2: ```suggestion
This however is a cause of concern because if the program tries to dereference and manipulate a pointer that holds no memory address, it could lead to some unexpected behavior. To prevent this, `nullptr` is used, which is a special kind of value that prevents our pointers from behaving unexpectedly when they hold no memory address.
```

```cpp

int *ptr = nullptr; // assings a null pointer value to our pointer

std::cout << *ptr << "\n"; // 0

std::string game = "Cricket";
*ptr = &game;

std::cout << *ptr << "\n"; // Cricket

```

> ## Pointers as Function Parameters

Reviewer 3: AP-style title case
```suggestion
## Pointers As Function Parameters
```

> We can also pass pointers into functions as arguments. This helps us manipulate values from within a function even though the value lives outside that function

Reviewer 2: ```suggestion
Pointers can also be passed into functions as arguments. This helps to manipulate values from within a function even though the value lives outside that function
```

```cpp

void changeGame(std::string *game, std::string new_game) { // defining a pointer as a function parameter

>   *game = new_game; // derefernce the pointer and change the game

Reviewer 3: ```suggestion
  *game = new_game; // Dereference the pointer and change the game
```

}

std::string game = "Cricket";

> changeGame(&game, "Baseball"); // passing a pointer as a function argument

Reviewer 3: ```suggestion
changeGame(&game, "Baseball"); // Passing a pointer as a function argument
```
